services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.linkedin.kmf.services;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import kafka.admin.AdminUtils;
+import kafka.admin.PreferredReplicaLeaderElectionCommand;
+import kafka.cluster.Broker;
+import kafka.common.TopicAndPartition;
+import kafka.utils.ZkUtils;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.security.JaasUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kmf.common.Utils.ZK_CONNECTION_TIMEOUT_MS;
+import static com.linkedin.kmf.common.Utils.ZK_SESSION_TIMEOUT_MS;
+
+import scala.collection.Seq;
+
+
+/**
+ * Runs this periodically to rebalance the monitored topic across brokers and to reassign leaders to brokers so that the
+ * monitored topic is sampling all the brokers evenly.
+ */
+class TopicRebalancer implements Runnable {
```

Comment:
One concern is that we currently need to restart `ProduceService` if we increase number of partition for the topic. Otherwise ProduceService will still only produce to old partitions. I talked with SRE and they told me they typically expend Kafka cluster once every quarter. The frequency of doubling the size of Kafka cluster will be even lower. They are OK with manually restarting Kafka Monitor every time they increase the size of Kafka cluster. Therefore we can make the code simpler by not having to restart ProduceService.